source/slang/slang-check-stmt.cpp

```diff
@@ -239,17 +239,57 @@ namespace Slang
         subContext.checkStmt(stmt->body);
     }
 
+    SeqStmt* SemanticsStmtVisitor::validateCaseStmts(SwitchStmt* stmt, DiagnosticSink* sink)
+    {
+        auto blockStmt = as<BlockStmt>(stmt->body);
+        if (!blockStmt)
+            return nullptr;
+
+        auto seqStmt = as<SeqStmt>(blockStmt->body);
+        if (!seqStmt)
+            return nullptr;
+
+        bool hasDefaultStmt = false;
+        HashSet<Val*> caseStmtVals;
+        for (auto& sStmt : seqStmt->stmts)
+        {
+            // check that all case tags are unique
+            if (auto caseStmt = as<CaseStmt>(sStmt))
+            {
+                if (auto caseVal = tryConstantFoldExpr(caseStmt->expr, ConstantFoldingKind::CompileTime, nullptr))
+                {
+                    if (caseStmtVals.contains(caseVal))
+                    {
+                        sink->diagnose(sStmt, Diagnostics::duplicateCasesInASwitch);
+                        return nullptr;
+                    }
+                    caseStmtVals.add(caseVal);
+                }
+            }
+            
+            // check that there is at most one `default` clause
+            else if (auto defaultStmt = as<DefaultStmt>(sStmt))
+            {
+                if (hasDefaultStmt)
+                {
+                    sink->diagnose(sStmt, Diagnostics::multipleDefaultsInASwitch);
+                    return nullptr;
+                }
+                hasDefaultStmt = true;
+            }
+        }
+        return seqStmt;
+    }
+
     void SemanticsStmtVisitor::visitSwitchStmt(SwitchStmt* stmt)
     {
         WithOuterStmt subContext(this, stmt);
 
         // TODO(tfoley): need to coerce condition to an integral type...
         stmt->condition = CheckExpr(stmt->condition);
-        subContext.checkStmt(stmt->body);
-
-        // TODO(tfoley): need to check that all case tags are unique
 
-        // TODO(tfoley): check that there is at most one `default` clause
+        auto seqStmt = validateCaseStmts(stmt, getSink());
```

Comment:
validate should happen after checking the stmt->body, so when you validate the case value exits, they are already type checked.

Reply:
ah i see. in that case, i see no reason to return the seqStmt, so i changed the validate to a void function.